bpf/packet_sniffer.c

```diff
@@ -144,21 +238,71 @@ static __always_inline int filter_packets(struct __sk_buff *skb, void *cgrpctxma
     __u16 src_port = 0;
     __u32 dst_ip = 0;
     __u16 dst_port = 0;
-    int ret = parse_packet(skb, 0, &src_ip, &src_port, &dst_ip, &dst_port, NULL);
+    struct in6_addr src_ip6 = {0};
+    struct in6_addr dst_ip6 = {0};
+    __u8 transportHdr = 0;
+    __u32 transportOffset = 0;
+    int ret = -1;
+
+    __u8 ip_version = 0;
+
+    if (bpf_skb_load_bytes(skb, 0, &ip_version, 1) < 0) {
+        return 1;
+    }
+
+    ip_version = (ip_version >> 4) & 0xF;
+
+    if (ip_version == 6) {
+        ret = parse_packet(skb, &src_ip, &src_port, &dst_ip, &dst_port, &transportHdr, &src_ip6, &dst_ip6, &transportOffset);
+    } else {
+        ret = parse_packet(skb, &src_ip, &src_port, &dst_ip, &dst_port, NULL, &src_ip6, &dst_ip6, NULL);
+    }
     if (!ret)
     {
         return 1;
     }
 
+    struct save_packet_args save_args = {
+        .skb = skb,
+        .cgroup_id = cgroup_id,
+        .direction = side,
+        .src_port = ( side == PACKET_DIRECTION_RECEIVED ? skb->remote_port>>16 : bpf_htons(skb->local_port)),
+        .dst_port = ( side == PACKET_DIRECTION_RECEIVED ? bpf_htons(skb->local_port) : skb->remote_port>>16),
+        .transportHdr = transportHdr,
+        .transportOffset = transportOffset,
+    };
+
     if (side == PACKET_DIRECTION_RECEIVED)
     {
-        TRACE_PACKET("cg/in", true, skb->local_ip4, skb->remote_ip4, skb->local_port & 0xffff, skb->remote_port & 0xffff, cgroup_id);
-        save_packet(skb, src_ip, skb->remote_port >> 16, dst_ip, bpf_htons(skb->local_port), cgroup_id, side);
+        if (src_ip) {
+            // IPv4
+            TRACE_PACKET_IPV4("cg/in", true, skb->local_ip4, skb->remote_ip4, skb->local_port & 0xffff, skb->remote_port & 0xffff, cgroup_id);
+            save_args.ipv4.src_ip = src_ip;
+            save_args.ipv4.dst_ip = dst_ip;
+            save_packet(&save_args);
+        } else {
+            // IPv6
+            TRACE_PACKET_IPV6("cg/in", true, skb->local_ip6, skb->remote_ip6, skb->local_port & 0xffff, skb->remote_port & 0xffff, cgroup_id);
+            save_args.ipv6.src_ip6 = src_ip6;
```

Comment:
It looks like saving pointers instead of copying whole IP, Isn't it?

I expected to see `memcpy` here